### DISCUSSION TITLE
Is the difference between the “worksheet” mode (`.sc` scripts) and “program” mode (`.scala` scripts) relevant?

### DISCUSSION THREAD

**Discussion by julienrf** [VirtusLab/scala-cli#1668] [Scripting]
As I mentioned in a [discussion](https://contributors.scala-lang.org/t/sip-46-scala-cli-as-default-scala-command/5996/13), there are currently two forms of scripts supported by Scala CLI. The `.sc` scripts are evaluated in a worksheet-like mode, where top-level statements are allowed, whereas the `.scala` scripts are evaluated as JVM programs (with a proper main method).

It is worth noting that the current `scala` command (based on Scala 2.13) does support both modes in a more “unified” way: regardless of the file extension, if the file defines a proper main method it is executed as a program, otherwise we fall back to a worksheet-like mode.

Do you think it would be possible (and desirable) to follow the same approach in Scala CLI? That could also help supporting scripts with no file extension at all (see https://github.com/VirtusLab/scala-cli/issues/466)?

**Comment by tgodzik** [comment]
That's an interesting issue. I would be clearer for the users if they wouldn't need a special extension. I had a user confused for a while when they were trying to execute `*.scala` file with top level statements. On the other hand, no tools actually support that aside from the runner and it might actually be more difficult to implement for us. IT would be simpler for the users for sure.

**Comment by julienrf** [comment]
I have just noticed that what i said about the “current” runner only applies to Scala 2, not Scala 3:

~~~ text
$ cat test.scala
println(s"Hello ${args(0)}")

$ cs launch scala:2.13.10 -- test.scala Julien
Hello Julien

$ cs launch scala:3.2.1 -- test.scala Julien
-- [E103] Syntax Error: /tmp/test.scala:4:0 ------------------------------------
4 |println(s"Hello ${args(0)}")
  |^^^^^^^
  |Illegal start of toplevel definition
  |
  | longer explanation available when compiling with `-explain`
1 error found
Errors encountered during compilation
~~~

**Comment by lwronski** [comment]
Yes, there is an relevant difference between `.sc` and `.scala` files.

For `.sc` files, `scala-cli` automatically adds the user code into the main method - I described this mechanism in more detail in https://github.com/VirtusLab/scala-cli/issues/1675#issuecomment-1351329254.

For `.scala` files treats as a JVM program with the proper main method, without checking the code inside.

@julienrf  Are you suggesting that `scala-cli` should check if the `.scala` file contains the correct main method and if so, run it. Otherwise put this code into the main method?  If we do this, integration with IDEE will be much more difficult.

**Reply by julienrf** [reply]
> Are you suggesting that `scala-cli` should check if the `.scala` file contains the correct main method and if so, run it. Otherwise put this code into the main method?

This is what the current Scala 2 runner does: regardless of the file extension, if the file defines a proper main method it is executed as a program, otherwise we fall back to a worksheet-like mode.

I wonder if the same behavior could be replicated here. That would also solve #466. But we should take into account the pros and cons. Maybe there is a better solution that would reduce the implementation costs.
